### ISSUE TITLE
Discount code expired can still be applied

### ISSUE THREAD

**Issue by chrisvidal** [OFFLINE-GmbH/oc-mall-plugin#401]
Discount code expired can still be applied on a cart and order.

**Comment by chrisvidal** [comment]
```
public function applyDiscountByCode(string $code)
    {
        $code = strtoupper(trim($code));
        if ($code === '') {
            throw new ValidationException([
                'code' => trans('offline.mall::lang.discounts.validation.empty'),
            ]);
        }

        try {
            $discount = Discount::whereCode($code)->firstOrFail();
        } catch (ModelNotFoundException $e) {
            throw new ValidationException([
                'code' => trans('offline.mall::lang.discounts.validation.not_found'),
            ]);
        }

        return $this->applyDiscount($discount);
    }
```

should be replaced with 

public function applyDiscountByCode(string $code)
    {
        $code = strtoupper(trim($code));
        if ($code === '') {
            throw new ValidationException([
                'code' => trans('offline.mall::lang.discounts.validation.empty'),
            ]);
        }

        try {
            $discount = Discount::whereCode($code)->
             where(function ($q) {
                $q->whereNull('valid_from')
                    ->orWhere('valid_from', '<=', Carbon::now());
            })->where(function ($q) {
                $q->whereNull('expires')
                    ->orWhere('expires', '>', Carbon::now());
            })->firstOrFail();

        } catch (ModelNotFoundException $e) {
            throw new ValidationException([
                'code' => trans('offline.mall::lang.discounts.validation.not_found'),
            ]);
        }

        return $this->applyDiscount($discount);
    }

**Comment by tobias-kuendig** [comment]
Fixed in 00f99ac, thanks again @chrisvidal!
